### PR TITLE
Logic to generate default cluster_name if not supplied

### PR DESCRIPTION
**Pull request by mitali-salvi**
*Description of changes:*
1. Implement logic to generate a default cluster-name (generated using the release name) if not supplied in Helm chart. This cluster name will be passed down to the cloudwatch-agent and fluent-bit config
2. Marking region as a required field. this region will be passed down to both the CWA and fluent-bit config

*Testing*
cluster_name is not provided, attaching snippets rather than the entire YAML
```
CWA --
  image: public.ecr.aws/cloudwatch-agent/cloudwatch-agent:1.300031.1b317
  mode: daemonset
  serviceAccount: cloudwatch-agent
  config: "{\"agent\":{\"region\":\"us-east-1\"},\"logs\":{\"metrics_collected\":{\"app_signals\":{\"hosted_in\":\"k8s-cluster-0ba3ba5\"},\"kubernetes\":{\"cluster_name\":\"k8s-cluster-0ba3ba5\",\"enhanced_container_insights\":true}}},\"traces\":{\"traces_collected\":{\"app_signals\":{}}}}"

FluentBit --
        image: public.ecr.aws/aws-observability/aws-for-fluent-bit:2.31.12.20230911
        imagePullPolicy: Always
        env:
        - name: AWS_REGION
          value: us-east-1
        - name: CLUSTER_NAME
          value: k8s-cluster-0ba3ba5
```

cluster_name is provided via the helm chart, attaching snippets rather than the entire YAML
```
helm upgrade --install  --debug --namespace amazon-cloudwatch amazon-cloudwatch-operator ./ --create-namespace --set clusterName=cloudwatchagent-operator --set region=us-east-1


CWA --
  image: public.ecr.aws/cloudwatch-agent/cloudwatch-agent:1.300031.1b317
  mode: daemonset
  serviceAccount: cloudwatch-agent
  config: "{\"logs\":{\"metrics_collected\":{\"app_signals\":{\"hosted_in\":\"k8s-cluster-0ba3ba5\"},\"kubernetes\":{\"cluster_name\":\"k8s-cluster-0ba3ba5\",\"enhanced_container_insights\":true}}},\"traces\":{\"traces_collected\":{\"app_signals\":{}}}}"

FluentBit --
        image: public.ecr.aws/aws-observability/aws-for-fluent-bit:2.31.12.20230911
        imagePullPolicy: Always
        env:
        - name: AWS_REGION
          value: us-east-1
        - name: CLUSTER_NAME
          value: "k8s-cluster-0ba3ba5"
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: helm/templates/_helpers.tpl [hunk 1]
```diff
@@ -5,6 +5,59 @@ Expand the name of the chart.
 {{- default .Chart.Name .Values.nameOverride | trunc 63 | trimSuffix "-" }}
 {{- end }}
 
+{{/*
+Name of k8s cluster
+*/}}
+{{- define "kubernetes-cluster.name" -}}
+{{- if eq .Values.clusterName "EKS_CLUSTER_NAME" }}
```
[the remark this answers is not in the document]
Comment (lisguo) [reply helm/templates/_helpers.tpl:12]: I think the intent is that if it equals the hardcoded string "EKS_CLUSTER_NAME", then it is not on EKS. Therefore we can create a generated cluster name

File: helm/templates/_helpers.tpl [hunk 2]
```diff
@@ -56,7 +116,7 @@ Common labels
 {{- if .Chart.AppVersion }}
 app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
 {{- end }}
-app.kubernetes.io/managed-by: EKS
+app.kubernetes.io/managed-by: "AmazonCloudWatchAgentOperator"
```
Comment (lisguo) [inline helm/templates/_helpers.tpl:119]: nit: amazon-cloudwatch-agent-operator

File: helm/values.yaml [hunk 3]
```diff
@@ -137,6 +137,8 @@ agent:
   config: # optional config that can be provided to override the defaultConfig
   defaultConfig:
     {
+      "agent": {
```
Comment (lisguo) [inline helm/values.yaml:140]: nit: can prob remove the agent section entirely

**Review by lisguo — DISMISSED** [review]

File: helm/templates/_helpers.tpl (hunk 2 above)
Comment (mitali-salvi) [reply helm/templates/_helpers.tpl]: updated

File: helm/values.yaml (hunk 3 above)
Comment (mitali-salvi) [reply helm/values.yaml]: removed and moved the logic to the helper function

**Review by lisguo — APPROVED** [review]

**Review by lisguo — APPROVED** [review]

**Review by movence — APPROVED** [review]

File: helm/templates/fluent-bit-daemonset.yaml [hunk 4]
```diff
@@ -28,7 +31,7 @@ spec:
         - name: AWS_REGION
           value: {{ .Values.region }}
         - name: CLUSTER_NAME
-          value: {{ .Values.clusterName | quote }}
+          value: {{ include "kubernetes-cluster.name" . }}
```
Comment (sky333999) [inline helm/templates/fluent-bit-daemonset.yaml:34]: @movence doesnt this bring back the bug you fixed for numeric only cluster names not being quoted?

File: helm/templates/_helpers.tpl [hunk 5]
```diff
@@ -5,6 +5,66 @@ Expand the name of the chart.
...
+{{- define "kubernetes-cluster.name" -}}
+{{- if empty .Values.clusterName }}
+{{- default "" (printf "k8s-cluster-%s" (sha256sum .Release.Name | trunc 7)) }}
+{{- else }}
+{{- default "" .Values.clusterName }}
+{{- end }}
+{{- end }}
+
+{{/*
+Helper function to modify cloudwatch-agent config
```
Comment (sky333999) [inline helm/templates/_helpers.tpl:20]: Whats diff between the two funcs below? 
I would have expected a generic helper function that takes a config (either user provided or default) as an input and then just merges region & cluster name on top of that.
```
.Values.agent.config | injectRegionAndClusterName | toJson | quote
.Values.agent.defaultConfig | injectRegionAndClusterName | toJson | quote
```

File: helm/values.yaml [hunk 6]
```diff
@@ -11,11 +11,11 @@ nameOverride: ""
...
 imagePullSecrets: [ ]
 
-## Provide the ClusterName (optional when installing via EKS add-on)
-clusterName: EKS_CLUSTER_NAME
+## Provide the ClusterName
+clusterName:
 
-## Provide the Region (optional when installing via EKS add-on)
-region: AWS_REGION_NAME
+## Provide the Region
```
Comment (sky333999) [inline helm/values.yaml:17]: Can we keep the comments on these saying this is required and the cluster name is optional and we generate one if not provdied

File: helm/templates/_helpers.tpl [hunk 7]
```diff
@@ -5,6 +5,66 @@ Expand the name of the chart.
 {{- default .Chart.Name .Values.nameOverride | trunc 63 | trimSuffix "-" }}
 {{- end }}
 
+{{/*
+Name of k8s cluster
+*/}}
+{{- define "kubernetes-cluster.name" -}}
+{{- if empty .Values.clusterName }}
+{{- default "" (printf "k8s-cluster-%s" (sha256sum .Release.Name | trunc 7)) }}
```
Comment (sky333999) [inline helm/templates/_helpers.tpl:13]: Is the release name unique enough for our purposes here? Is a hash good enough or do we need randomness?
If the same release is installed on multiple clusters with the same name, wont all the clusters have the same hash?

Comment (movence) [reply helm/templates/_helpers.tpl:13]: This is a good point, and we should add some randomness to it.

**Pull request merged**